Library/Homebrew/unpack_strategy/dmg.rb

```diff
@@ -87,10 +87,31 @@ def eject(verbose: false)
         return unless path.exist?
 
         if tries > 1
-          system_command! "diskutil",
-                          args:         ["eject", path],
-                          print_stderr: false,
-                          verbose:      verbose
+          disk_info = system_command!(
+            "diskutil",
+            args:         ["info", "-plist", path],
+            print_stderr: false,
+            verbose:      verbose,
+          )
+
+          # For HFS, just use <mount-path>
+          # For APFS, find the <physical-store> corresponding to <mount-path>
+          eject_paths = if disk_info.success?
+            disk_info.plist
+                     .fetch("APFSPhysicalStores", [])
+                     .map { |store| store["APFSPhysicalStore"] }
+                     .compact
+                     .presence || [path]
+          else
+            [path]
+          end
+
+          eject_paths.each do |eject_path|
+            system_command! "diskutil",
+                            args:         ["eject", eject_path],
+                            print_stderr: false,
+                            verbose:      verbose
+          end
         else
           system_command! "diskutil",
                           args:         ["unmount", "force", path],
```

Comment:
This has to be `unmountDisk` in the APFS case since `APFSPhysicalStore` isn't a path but a disk ID.

Maybe use the `WholeDisks` entry for HFS to also map the path to a disk.

Reply:
Both `unmount` and `unmountDisk` do work with diskID, path, etc.
I believe it converts the diskID to volume paths.

`unmount` actually is a no-op on APFS physical store since there is no Volumes on there.
It either needs the virtual disk partition or volume path, e.g.
```console
/dev/disk4 (disk image):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:      GUID_partition_scheme                        +436.7 MB   disk4
   1:                 Apple_APFS Container disk5         436.6 MB   disk4s1

/dev/disk5 (synthesized):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:      APFS Container Scheme -                      +436.6 MB   disk5
                                 Physical Store disk4s1
   1:                APFS Volume calibre-5.14.0          400.4 MB   disk5s1

❯ diskutil unmount disk4s1
disk4s1 was already unmounted

❯ diskutil unmount disk5s1
Volume calibre-5.14.0 on disk5s1 unmounted
```

`unmountDisk` is more like loop over all related disks, so can be run on APFS physical or virtual and should behave identically:
```console
❯ ls /Volumes
'Macintosh HD'   calibre-5.14.0

❯ diskutil unmountDisk force disk4s1
Forced unmount of all volumes on disk4 was successful

❯ ls /Volumes
'Macintosh HD'

❯ diskutil list disk4
/dev/disk4 (disk image):
   #:                       TYPE NAME                    SIZE       IDENTIFIER
   0:      GUID_partition_scheme                        +436.7 MB   disk4
   1:                 Apple_APFS Container disk5         436.6 MB   disk4s1
```

Reply:
Also, probably should leave `unmount` since the original HFS logic was designed to create a Mount object for every volume path:
```ruby
          plist.fetch("system-entities", [])
               .map { |entity| entity["mount-point"] }
```

Eject logic is called for each mount:
```ruby
          mounts.each do |mount|
            mount.eject(verbose: verbose)
          end
```

This is probably programmatically similar to `unmountDisk`.

New APFS logic is fine due to check whether Volume path exists prevents calling eject/unmount on nonexistent device.

Reply:
Neither `unmount force` nor `unmountDisk force` seem to work for the APFS physical disk.

```
diskutil unmountDisk force /dev/disk2s1
```

says

```
Forced unmount of all volumes on disk2 was successful
```

yet `disk2` is still mounted.

Reply:
I think the unmount only removes the mounted Volumes.
e.g. If you do standard mount into `/Volumes/...`, it should no longer be mounted in that directory.

`diskutil eject` is required to remove/clean-up the actual link to node `/dev/diskX`.

I think HFS behaves the same as APFS here.
Both should unmount the Volumes but leave behind the `/dev/diskX` entry in the case that the fallback scenario happens and `diskutil unmount force` is used.

It isn't ideal, but I guess partial ejection (i.e. only unmounting) is better than nothing.